File: app/src/main/java/ws/grigory/currencycalculator/calculator/Calculator.java
package ws.grigory.currencycalculator.calculator;

import static ws.grigory.currencycalculator.CCWidgetLarge.C;
import static ws.grigory.currencycalculator.CCWidgetLarge.DIV;
import static ws.grigory.currencycalculator.CCWidgetLarge.EVAL;
import static ws.grigory.currencycalculator.CCWidgetLarge.MINUS;
import static ws.grigory.currencycalculator.CCWidgetLarge.MUL;
import static ws.grigory.currencycalculator.CCWidgetLarge.PLUS;
import static ws.grigory.currencycalculator.CCWidgetLarge.SHIFT;
import static ws.grigory.currencycalculator.CCWidgetLarge.ZERO_CHAR;
import static ws.grigory.currencycalculator.calculator.Display.EMPTY;

import android.content.Context;

import java.util.ArrayList;

import ws.grigory.currencycalculator.Currency;
import ws.grigory.currencycalculator.WidgetParameters;

public class Calculator {
    private static final String OPERATIONS = String.valueOf(new char[]{PLUS, MINUS, MUL, DIV, EVAL});
    private float yRegistry = 0;
    private char operation = EVAL;
    private final Display[] displays = new Display[3];
    private MainDisplay mainDisplay;
    private boolean isNewValue = true;
    private ArrayList<Currency> currencies;

    public Calculator(Context context) {
        loadCurrencies(context);
        mainDisplay.processDisplayChar(true, ZERO_CHAR);
    }

    public void setData(char data, Context context) {
        Currency currency = mainDisplay.getMainDisplayCurrency();
        if (C == data) {
            clear();
        } else {
            if (!mainDisplay.getMainDisplayCurrency().isInfinity()) {
                if (SHIFT == data) {
                    shift(context);
                } else {
                    if (OPERATIONS.indexOf(data) < 0) {
                        isNewValue = mainDisplay.processDisplayChar(isNewValue, data);
                        mainDisplay.writeDisplayValueToCurrency();
                    } else {
                        if (!isNewValue) {
                            switch (operation) {
                                case PLUS:
                                    currency.plus(yRegistry);
                                    break;
                                case MINUS:
                                    currency.minus(yRegistry);
                                    break;
                                case MUL:
                                    currency.mul(yRegistry);
                                    break;
                                case DIV:
                                    currency.div(yRegistry);
                            }
                            mainDisplay.writeCurrencyValueToDisplay();
                        }

                        if (currency.isInfinity() || data == EVAL) {
                            clearRegistries();
                        } else {
                            operation = data;
                            yRegistry = currency.value;
                        }
                        isNewValue = true;
                    }
                    mainDisplay.createExpression(operation, yRegistry);
                    calculateCurrencies(currency.value);
                }
            }
        }
    }

    private void calculateCurrencies(float value) {
        mainDisplay.getMainDisplayCurrency().value = value;
        float baseValue = mainDisplay.getMainDisplayCurrency().getBaseValue();

        for (int i = 1; i < displays.length; i++) {
            displays[i].calculateCurrencyValueToDisplay(baseValue);
        }
    }

    private void shift(Context context) {
        for (Display display : displays) {
            display.shift();
        }
        clearRegistries();
        WidgetParameters.getWidgetParameters(context).
                saveParameters(context, currencies, displays[0].currencyIndex);
    }

    public void clear() {
        clearRegistries();

        for (Display display : displays) {
            display.zeroingDisplay();
        }

        mainDisplay.processDisplayChar(true, ZERO_CHAR);
        mainDisplay.createExpression(operation, yRegistry);
    }

    public void reset(Context context){
        loadCurrencies(context);
        displays[0].currencyIndex = 0;
        displays[0].currencies = currencies;
        displays[1].currencyIndex = 1;
        displays[1].currencies = currencies;
        displays[2].currencyIndex = 2;
        displays[2].currencies = currencies;
        clear();
    }

    private void clearRegistries() {
        yRegistry = 0;
        operation = EVAL;
        isNewValue = true;
    }

    public Display[] getDisplays() {
        return displays;
    }

    @SuppressWarnings("unchecked")
    private void loadCurrencies(Context context) {
        WidgetParameters widgetParameters = WidgetParameters.getWidgetParameters(context);
        ArrayList<Currency> data = widgetParameters.currencies;
        if(data == null) {
            currencies = new ArrayList<>(3);
            currencies.add(new Currency(EMPTY));
        } else {
            currencies = (ArrayList<Currency>)data.clone();
        }
        int currencyIndex = widgetParameters.mainDisplayCurrencyIndex;
        mainDisplay = new MainDisplay(currencyIndex, this.currencies);
        displays[0] = mainDisplay;
        currencyIndex = currencyIndex == 0 ? currencies.size() - 1 : --currencyIndex;
        displays[1] = new Display(currencyIndex, this.currencies);
        currencyIndex = currencyIndex == 0 ? currencies.size() - 1 : --currencyIndex;
        displays[2] = new Display(currencyIndex, this.currencies);

    }
}

File: app/src/main/java/ws/grigory/currencycalculator/calculator/Calculator.kt
package ws.grigory.currencycalculator.calculator

import android.content.Context
import ws.grigory.currencycalculator.Constants.C
import ws.grigory.currencycalculator.Constants.DIV
import ws.grigory.currencycalculator.Constants.EVAL
import ws.grigory.currencycalculator.Constants.MINUS
import ws.grigory.currencycalculator.Constants.MUL
import ws.grigory.currencycalculator.Constants.PLUS
import ws.grigory.currencycalculator.Constants.SHIFT
import ws.grigory.currencycalculator.Constants.ZERO_CHAR

import ws.grigory.currencycalculator.Currency
import ws.grigory.currencycalculator.WidgetParameters

class Calculator(context: Context) {
    private val operations = charArrayOf(PLUS, MINUS, MUL, DIV, EVAL).toHashSet()
    private var yRegistry: Float = 0F
    private var operation: Char = EVAL
    private var isNewValue: Boolean = true
    private lateinit var displays: Array<Display>
    private lateinit var mainDisplay: MainDisplay
    private lateinit var currencies: ArrayList<Currency>
    var invalidated = true

    init {
        loadCurrencies(context)
        mainDisplay.processDisplayChar(true, ZERO_CHAR)
        invalidated = true
    }

    fun setData(data: Char, context: Context) {
        invalidated = false
        val currency: Currency = mainDisplay.mainDisplayCurrency
        if (C == data) {
            clear()
        } else {
            if (!mainDisplay.mainDisplayCurrency.isInfinity()) {
                if (SHIFT == data) {
                    shift(context)
                } else {
                    if (!operations.contains(data)) {
                        isNewValue = mainDisplay.processDisplayChar(isNewValue, data)
                        mainDisplay.writeDisplayValueToCurrency()
                    } else {
                        if (!isNewValue) {
                            when (operation) {
                                PLUS -> currency.plus(yRegistry)
                                MINUS -> currency.minus(yRegistry)
                                MUL -> currency.mul(yRegistry)
                                DIV -> currency.div(yRegistry)
                            }
                            mainDisplay.writeCurrencyValueToDisplay()
                        }

                        if (currency.isInfinity() || data == EVAL) {
                            clearRegistries()
                        } else {
                            operation = data
                            yRegistry = currency.value
                        }
                        isNewValue = true
                    }
                    mainDisplay.createExpression(operation, yRegistry)
                    calculateCurrencies(currency.value)
                }
            }
        }
    }

    private fun calculateCurrencies(value: Float) {
        mainDisplay.mainDisplayCurrency.value = value
        val baseValue: Float = mainDisplay.mainDisplayCurrency.getBaseValue()
        displays.filter { display: Display -> display !is MainDisplay }
            .forEach { display -> display.calculateCurrencyValueToDisplay(baseValue) }
    }

    private fun shift(context: Context) {
        displays.forEach { display -> display.shift() }
        clearRegistries()
        WidgetParameters.getWidgetParameters(context).saveParameters(
            context, currencies, displays[0].currencyIndex
        )
    }

    fun reset(context: Context) {
        loadCurrencies(context)
        displays.forEachIndexed { index, display ->
            run {
                display.currencyIndex = index
                display.currencies = currencies
            }
        }
        clear()
    }

    private fun clear() {
        clearRegistries()
        displays.forEach { display -> display.zeroingDisplay() }

        mainDisplay.processDisplayChar(true, ZERO_CHAR)
        mainDisplay.createExpression(operation, yRegistry)
    }

    private fun clearRegistries() {
        yRegistry = 0F
        operation = EVAL
        isNewValue = true
        invalidated = true
    }

    fun getDisplays(): Array<Display> {
        return displays
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadCurrencies(context: Context) {
        val widgetParameters: WidgetParameters = WidgetParameters.getWidgetParameters(context)

        currencies = widgetParameters.currencies.clone() as ArrayList<Currency>

        mainDisplay = MainDisplay(widgetParameters.mainDisplayCurrencyIndex, currencies)

        displays = arrayOf(
            mainDisplay,
            Display(getCurrencyIndex(1, widgetParameters.mainDisplayCurrencyIndex), currencies),
            Display(getCurrencyIndex(2, widgetParameters.mainDisplayCurrencyIndex), currencies)
        )
    }

    fun invalidate() {
        clearRegistries()
        mainDisplay.createExpression(EVAL, 0F)
    }

    private fun getCurrencyIndex(order: Int, currencyIndex: Int): Int {
        return if (currencies.size > order)
            if (currencyIndex == 0) order
            else if (currencies.size == order + currencyIndex) 0
            else currencies.size - currencyIndex
        else
            -1
    }
}
